### PR TITLE
chore: Move public-api-check script to GH action

### PR DESCRIPTION
**Pull request by shibeshduw**
The check public-api-script is converted to a GH action with inputs. This is then reused in the AI SDK ([PR](https://github.com/SAP/ai-sdk-js/pull/254)).
Closes https://github.tools.sap/AI/gen-ai-hub-sdk-js-backlog/issues/115

### CODE REVIEW

**Comment by shibeshduw** [comment]
**NOTE**

Ignore the many junk commits. I couldn't get [nektos/act](https://github.com/nektos/act) to run actions locally on my machine, so I had to do some trial and error here.

The public api check works as intended both here and in the [AI SDK](https://github.com/SAP/ai-sdk-js/pull/254). 
The only problem here is with the unit tests for the public api check. [This test case](https://github.com/SAP/cloud-sdk-js/blob/e47cc3e38fa283ad8250e9c62ef54e2328205c48/build-packages/check-public-api/check-public-api.spec.ts#L58) passes in one `build/tests` action but fails in the other two `build/tests` actions (see the checks above). According to the current logic, the current code is correct and the test should pass in every case. The issue arises with `exportAllInBarrel()` not identifying the barrelFile that has been mocked in the test.

File: build-packages/check-public-api/check-public-api.spec.ts [hunk 1]
```diff
@@ -59,64 +65,64 @@ describe('check-public-api', () => {
...
+      ]);
     });
-    expect(typeDescriptorPaths('dir1')).toEqual([
-      'dir1/dir2/file2.d.ts',
-      'dir1/dir2/file3.d.ts',
-      'dir1/file1.d.ts'
-    ]);
   });
 
-  describe('parseTypeDefinitionFile', () => {
```
Comment (deekshas8) [inline build-packages/check-public-api/check-public-api.spec.ts:106]: [q] Not sure if i understood what changed here, did we rename the function?

**Review by deekshas8 — COMMENTED** [review]
It seems existsSync doesn't work consistently between node 18 and 20/22.

File: build-packages/check-public-api/check-public-api.spec.ts [hunk 2]
```diff
@@ -1,52 +1,58 @@
...
+      });
+      checkIndexFileExists('root/index.ts');
+      expect(errorSpy).toHaveBeenCalledWith('No index.ts file found in root.');
+    });
 
     it('fails if internal.ts is not present in root', async () => {
       mock({
         src: {
           file1: '',
+          'internal.ts': '',
```
Comment (deekshas8) [inline build-packages/check-public-api/check-public-api.spec.ts:43]: After adding this line, the test fails in 18 (as it should) but still passes in 20/22. So I suspect the behavior is different. Maybe something to look into

File: build-packages/check-public-api/check-public-api.spec.ts (hunk 1 above)
Comment (shibeshduw) [reply build-packages/check-public-api/check-public-api.spec.ts:106]: Yep. Just a name change. The current name makes more sense than what is was.

File: package.json [hunk 3]
```diff
@@ -68,7 +69,7 @@
     "check:dependencies": "turbo run check:dependencies",
     "check:license": "node ./.github/actions/check-license/index.js",
     "check:test-service": "yarn generate && yarn ts-node scripts/check-test-service-for-changes.ts",
-    "check:public-api": "turbo run check:public-api",
+    "check:public-api": "node ./.github/actions/check-public-api/index.js",
```
[the remark this answers is not in the document]
Comment (shibeshduw) [reply package.json:72]: Yes. Overlooked this. Thanks for reminding.

[the remark this answers is not in the document]
Comment (shibeshduw) [reply package.json:72]: Turns out there is no way of using `getInput()` locally. So I've created a local-config.json and conditionally load the options from there when it cannot be found using `getInput`.

File: build-packages/check-public-api/check-public-api.spec.ts (hunk 2 above)
Comment (shibeshduw) [reply build-packages/check-public-api/check-public-api.spec.ts]: I've used a workaround with `accessSync` instead. Found a small section on [iojs.org](https://iojs.org/api/fs.html#fs_fs_access_path_mode_callback:~:text=This%20is%20useful%20for%20determining%20if%20a%20file%20exists) that suggests this as an alternative. 
Works well and all the tests pass as normal now.

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (deekshas8) [reply package.json:72]: Have you tried this https://stackoverflow.com/questions/63722151/set-github-actions-input-value-for-local-testing ?

[the remark this answers is not in the document]
Comment (shibeshduw) [reply package.json]: Ended up using the `cross-env` package for setting env variables. The reason being neither `set` nor `exports` are truly cross platform and don't work in either unix/windows.

[the remark this answers is not in the document]
Comment (deekshas8) [reply package.json]: Can you run the windows-tests workflow on this branch? Just to confirm.

[the remark this answers is not in the document]
Comment (deekshas8) [reply package.json]: Also, can we clean up the original script in the scripts folder?

[the remark this answers is not in the document]
Comment (shibeshduw) [reply package.json]: windows-test workflow [passes](https://github.com/SAP/cloud-sdk-js/actions/runs/11819917042). also cleaned up the original script.

**Review by deekshas8 — APPROVED** [review]
LGTM

**Pull request merged**